packages/core/src/database/index.ts

```diff
@@ -69,21 +72,51 @@ export type Database = {
   ) => Promise<T>;
   /** Migrate the `ponder_sync` schema. */
   migrateSync(): Promise<void>;
-  /** Migrate the user schema. */
+  /**
+   * Migrate the user schema.
+   *
+   * @returns The crash recovery checkpoint for each chain if there is a cache hit, else undefined.
+   */
   migrate({
     buildId,
-  }: Pick<IndexingBuild, "buildId">): Promise<string | undefined>;
+  }: Pick<IndexingBuild, "buildId">): Promise<CrashRecoveryCheckpoint>;
   createIndexes(): Promise<void>;
   createTriggers(): Promise<void>;
   removeTriggers(): Promise<void>;
-  getStatus: () => Promise<Status | null>;
-  setStatus: (status: Status) => Promise<void>;
+  /**
+   * - "safe" checkpoint: The closest-to-tip finalized and completed checkpoint.
+   * - "latest" checkpoint: The closest-to-tip completed checkpoint.
+   *
+   * @dev It is an invariant that every "latest" checkpoint is specific to that chain.
+   * In other words, `chainId === latestCheckpoint.chainId`.
+   */
+  setCheckpoints: ({
+    checkpoints,
+  }: {
+    checkpoints: {
+      chainName: string;
+      chainId: number;
+      safeCheckpoint: string;
+      latestCheckpoint: string;
+    }[];
+    db: Drizzle<Schema>;
```

Comment:
I understand why (sometimes in transaction, sometimes not) but noting that this looks a bit weird.

Reply:
I have plans to fix, but out of the scope of this pr.